docs: mdbook guide

Mirko-von-Leipzig opened the pull request:
This PR adds an `mdbook` doc guide for node operators and developers.

I've been staring at this for a bit too long; I think I've developed blind spots. Happy with any feedback.

Note: this includes a `mermaid.js` diagram - I'm not sure its that nice, there are maybe better formats to choose instead.

Closes #244

phklive commented on docs/src/SUMMARY.md:
```diff
@@ -0,0 +1,30 @@
+# Summary
+
+[Introduction](./introduction.md)
+
+---
+
+# Operator Guide
+
+- [Overview](./operator/overview.md)
+- [Versioning](./operator/versioning.md)
+- [Architecture](./operator/architecture.md)
+- [Installation](./operator/installation.md)
+- [Configuraton and Usage](./operator/configuration.md)
+- [Monitoring](./operator/monitoring.md)
+
+---
+
+# Developer Guide
+
+- [Overview](./developer/overview.md)
+- [Contributing](./developer/contributing.md)
+- [Navigating the codebase](./developer/codebase.md)
+- [Monitoring](./developer/monitoring.md)
+- [Components](./developer/components.md)
+    - [RPC](./developer/rpc.md)
+    - [Store](./developer/store.md)
+    - [Block producer](./developer/block-producer.md)
+- [Common issues other oddities](./developer/oddities.md)
+
+---
```
```suggestion
# Summary

- [Node](./introduction.md)
    - [Operator](./operator/overview.md)
        - [Versioning](./operator/versioning.md)
        - [Architecture](./operator/architecture.md)
        - [Installation](./operator/installation.md)
        - [Configuraton and Usage](./operator/configuration.md)
        - [Monitoring](./operator/monitoring.md)
    - [Developer](./developer/overview.md)
         - [Contributing](./developer/contributing.md)
         - [Navigating the codebase](./developer/codebase.md)
         - [Monitoring](./developer/monitoring.md)
         - [Components](./developer/components.md)
             - [RPC](./developer/rpc.md)
             - [Store](./developer/store.md)
             - [Block producer](./developer/block-producer.md)
         - [Common issues other oddities](./developer/oddities.md)
```

phklive replied:
That's the structure we need to be able to pull the into the Miden book, don't worry they look good in there.

Mirko-von-Leipzig replied:
Well no they're subsections within a single chapter as opposed to multiple chapters and multiple segments.

I imagine especially developers would be reading this book here not the general documentation one.

I'll make this change but it would be nice not to have to mutilate the one in favor of the other

phklive commented on docs/src/introduction.md:
```diff
@@ -0,0 +1,13 @@
+# Introduction
```
I would also add some information about:
- What the node is in the context of the Miden protocol
- How does it work
- etc...
Basically can you try to summarize everything that you have put in the docs in a one-pager. (Does not need to be very long)

phklive commented on docs/src/introduction.md:
```diff
@@ -0,0 +1,13 @@
+# Introduction
+
+Welcome to the Miden node documentation.
+
```
Would change name to `index.md`

phklive commented on docs/src/developer/overview.md:
```diff
@@ -0,0 +1,19 @@
+
+<div class="warning">
+
+Living documents go stale - the code is the final arbitrator of truth.
```
I would change the name of the file to `index.md`

phklive commented on docs/src/operator/overview.md:
```diff
@@ -0,0 +1,7 @@
+
+Welcome to the `Miden` node operator guide which should cover everything you need to succesfully run and maintain a
+Miden node.
+
```
I would change the name of the file to `index.md`

phklive reviewed (approved):
Great to see that, few changes in structure but overall LGTM.

Mirko-von-Leipzig commented:
I've updated the PR with some minor changes.

As discussed offline I'll leave this targeting `next` and open a new PR for `main` for the current v0.7 docs.

Mirko-von-Leipzig commented on .github/workflows/book.yml:
```diff
@@ -0,0 +1,94 @@
+    name: deploy mdbook
+    runs-on: ubuntu-latest
+    needs: check-files
+    # Only run deployment on push to main/next branches, not on PRs
+    if: github.event_name == 'push' || github.event_name == 'workflow_run'
```
This means pushes to `main` and `next` will overwrite each other. Probably we would want this to only happen on `next` so that they're always the latest dev docs.

Mirko-von-Leipzig reviewed (commented):
As I mentioned offline, I think we need a clearer picture of which docs we want where before we merge this.

bobbinth commented on docs/src/operator/architecture.md:
````diff
@@ -0,0 +1,65 @@
+The inter-component communication is done using a gRPC API wnich is assumed trusted. In other words this _must not_ be
+public. External communication is handled by the RPC component with a separate external-only gRPC API.
+
+```mermaid
+---
+title: Infrastructure architecture example
+---
+graph LR;
+    
+  subgraph RPC
+  direction RL
+    rpc_a[RPC A]
+    rpc_b[RPC B]
+    rpc_i[......]
+    rpc_n[RPC N]
+  end
+
+  subgraph internal
+  direction TB
+    store
+    block-producer
+  end
+
+  load_balancer[Load Balancer]
+
+  load_balancer ---> RPC
+
+  RPC -- tx   ---> block-producer 
+  RPC -- query --> store
+
+  block-producer -- block --> store
+```
````
I think a simple Excalidraw diagram (like we have in the readme) may be better here (we'd have more control of what to show and how). I've attached the source code for the diagrams I have.

[node design.excalidraw.zip](https://github.com/user-attachments/files/19153267/node.design.excalidraw.zip)

Mirko-von-Leipzig replied:
I've used your image here instead.

bobbinth commented on docs/src/operator/architecture.md:
```diff
@@ -0,0 +1,65 @@
+
+## Block-producer
+
+The block-producer is responsible for aggregating received transactions into blocks and submitting them to the store.
+
+Transactions are placed in a mempool and are periodically sampled to form batches of transactions. These batches are
+proved, and then periodically aggregated into a block. This block is then proved and committed to the store.
```
I think this was written before we merged the relevant PRs, but I'd also mention how proving works (e.g., that batch and block proving can be done via remote provers).

bobbinth commented on docs/src/operator/installation.md:
```diff
@@ -0,0 +1,85 @@
+# Installation
+
+We provide Debian packages for official releases for both the node software as well as a basic faucet implementation.
```
The mentioning of the faucet comes out a bit "out of the blue". We should probably have a small section somewhere explaining what the faucet is and why it is needed.

Mirko-von-Leipzig replied:
I've shuffled things around a bit, and I introduce the faucet in `# Architecture`.

bobbinth commented on docs/src/operator/installation.md:
```diff
@@ -0,0 +1,85 @@
+
+## Install using `cargo`
+
+Install Rust version **1.84** or greater using the official Rust installation
+[instructions](https://www.rust-lang.org/tools/install).
```
I think Rust version should now be 1.85 - right?

bobbinth commented on docs/src/operator/hardware.md:
Doesn't seem like this file is used anywhere.

Mirko-von-Leipzig replied:
Usually a question is: what hardware does the node need so I wanted to address this early.

We can't really answer that question yet - I don't think? In which case any answer might be a bad answer.

I basically wanted to indicate that you'll need extra stuff to run provers. But I'll just punt that problem and yank this file.

bobbinth commented on docs/src/operator/monitoring.md:
```diff
@@ -39,19 +15,22 @@ OpenTelemetry provides a
+</div>
 
-#### RPC request/response
+### RPC request/response
 
 Not yet implemented.
```
I thought this has already been implemented?

Mirko-von-Leipzig replied:
They are for the internal gRPC communication i.e. they're added as part of the block-producer's spans. However they haven't been added for requests arriving at the RPC component.

bobbinth commented on docs/src/developer/codebase.md:
````diff
@@ -0,0 +1,34 @@
+instead simply serve to enforce code organised and decoupling. The node binary is used to launch the components i.e. we
+have a single binary with subcommands for each component (and running them all as a single process).
+
+```mermaid
+---
+title: Workspace dependency tree
+---
+graph TD;
+    
+  subgraph components
+      block-producer
+      store
+      rpc
+  end
+
+  faucet ---> proto & utils;
+  node --> components & utils
+
+  components --> proto & utils
+
+  proto; utils;
+```
````
I would maybe just have a simple table with crate name and description - kind of similar to how we do it in [miden-vm](https://github.com/0xPolygonMiden/miden-vm?tab=readme-ov-file#project-structure) readme  and maybe complement it with an excalidraw diagram.

bobbinth commented on docs/src/developer/rpc.md:
Not for this PR, but maybe this is the place where we add the description of RPC component API (i.e., the endpoint docs).

Mirko-von-Leipzig replied:
I think we can have a separate `# RPC reference` section, which could also be part of the user guide?

bobbinth commented on docs/src/developer/store.md:
```diff
@@ -0,0 +1,19 @@
+on node startup to ensure that any existing database matches the expected schema. If you're seeing database failures on
+startup its likely that you created the database _before_ making schema changes resulting in different schema hashes.
+
+## Architecture
+
+The store consists mainly of a gRPC server which answers requests from the RPC and block-producer components, as well as
+new block submissions from the block-producer.
+
+A lightweight background process performs database query optimisation by analysing database queries and statistics.
```
Not for this PR, but I would maybe add more info about how the store works - e.g., that we have accounts, notes, and nullifiers; that Merkle data structures are currently always in memory (and are re-built on node startup).

Mirko-von-Leipzig replied:
I've tried to address this with a short paragraph.

bobbinth commented on docs/src/developer/block-producer.md:
I would mention somewhere in this file that batches/blocks can be proven by remote provers.

bobbinth commented on docs/src/IMPORTED.md:
```diff
@@ -0,0 +1,19 @@
+    - [Installation](./operator/installation.md)
+    - [Usage](./operator/usage.md)
+    - [Monitoring](./operator/monitoring.md)
+  - [Developer](./developer/index.md)
+    - [Contributing](./developer/contributing.md)
+    - [Navigating the codebase](./developer/codebase.md)
+    - [Monitoring](./developer/monitoring.md)
+    - [Components](./developer/components.md)
+      - [RPC](./developer/rpc.md)
+      - [Store](./developer/store.md)
+      - [Block producer](./developer/block-producer.md)
+    - [Common issues other oddities](./developer/oddities.md)
```
I'm not sure this section needs to go into the unified docs.

bobbinth commented on docs/src/IMPORTED.md:
nit: as mentioned in some other PRs, I would probably name this `EXPORTED.md`.

bobbinth commented on .github/workflows/book.yml:
As mentioned in the reviews in parallel PRs in other repos: I'd probably simplify this file to only keep the `deploy` job and make this job similar to what we currently have in `miden-vm` (unless there is a good reason to deviate).

bobbinth reviewed (commented):
Looks great! Thank you! I left some comments inline - most are pretty minor.

Once this is ready to merge, I'll create a Github page for the mdBook (similar to how we have it in the `miden-vm` repo) and then we'll need to update the main readme to link to this book.

Mirko-von-Leipzig commented:
@bobbinth @phklive I believe I've addressed all comments.

ito infrastructure it should be correct - though @phklive I'm unsure what the `EXPORTER.md` section should look like for you. Right now it only contains the operator guide so my index description isn't very valid for your purposes.

bobbinth commented on docs/book.toml:
```diff
@@ -0,0 +1,8 @@
+[book]
+authors      = ["Miden"]
```
nit: maybe "Miden team"?

bobbinth commented on docs/src/operator/architecture.md:
```diff
@@ -0,0 +1,47 @@
+The inter-component communication is done using a gRPC API wnich is assumed trusted. In other words this _must not_ be
+public. External communication is handled by the RPC component with a separate external-only gRPC API.
+
+![node architecture](../resources/operator_architecture.svg)
```
I know this is my picture - but I think a couple of things here are confusing:

- The purpose of the explorer box is not super clear here. I think what we try to show is that "explorer.testnet.miden.io" gets redirected to it - but this is not really obvious.
- The "faucet" box is also a bit confusing - it feels like the faucet is not really communicating with any other component - but I think it does talk to an internal RPC, right?

Mirko-von-Leipzig replied:
I've made some changes to the diagram:

- pass through arrows connecting the load balancer inputs to outputs
- marked the explorer as an external service
- added provers
- added missing links

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of small comments inline.

Mirko-von-Leipzig commented:
@phklive you'll likely need to open another PR to adjust the exporting, but I wanted to get the bulk of this in so current PRs can start updating the content.

bobbinth commented:
It seems like `mdbook` deploy action ran but the book didn't actually get deployed - i.e., https://0xpolygonmiden.github.io/miden-node/ (which redirects to `gh-pages` branch) results in a 404.
